### PR TITLE
Rearrange GDD tests to cut down runtime.

### PR DESCRIPTION
**Pull request by ashwinstar**
`prepare-for-local` test resets the `gp_global_deadlock_detector_period`
guc to default 2mins time. Due to it any GDD test after that will take
at least 2mins. Instead flip the order.

Ideally `prepare-for-local` test don't need to reset the guc. Instead
can use fault injector in local-deadlock-03 to avoid GDD from
kicking-in, but that's for separate commit.

This shaves off **6mins** of isolation2 test time on my laptop.

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by ashwinstar** [comment]
also planning to have `gp_global_deadlock_detector_period` as 5 sec instead of 10sec to cut down time even further in `prepare.sql`. Plus, also avoid `gpstop -rai` in `prepare.sql` and `end.sql` instead can just do `pg_ctl -D <master_data_directory> restart` to knock off somemore mins.

**Comment by pivotal-ning-yu** [comment]
To further reduce the runtime we could put the tests into a parallel group, we just need to ensure the tests are using different tables or schemas.  I had a quick check that only `planner_insert_while_vacuum_drop.sql` need a separate group due to fault injection, other tests should all be put into the same parallel group.

> also planning to have `gp_global_deadlock_detector_period` as 5 sec instead of 10sec to cut down time even further in `prepare.sql`. Plus, also avoid `gpstop -rai` in `prepare.sql` and `end.sql` instead can just do `pg_ctl -D <master_data_directory> restart` to knock off somemore mins.

Please do not reduce the period, sometimes the pipelines can be really slow, we need enough time to perform the checks in the tests.  An alternative way is to use fault injection to control when to suspend and resume gdd, but that might require the tests to be run one by one, which may not be a good deal.

**Comment by pivotal-ning-yu** [comment]
> > > Please do not reduce the period, sometimes the pipelines can be really slow, we need enough time to perform the checks in the tests. An alternative way is to use fault injection to control when to suspend and resume gdd, but that might require the tests to be run one by one, which may not be a good deal.
> 
> I am missing something here, why should reducing the `gp_global_deadlock_detector_period` affect the stability of the GDD check. It just controls interval for how much time GDD backend sleeps in-between checks. If pipelines are running slow, why checking more often affects the checks? I don't think fault injector is route to go here.
> 
> First I don't believe reducing `gp_global_deadlock_detector_period` affects checks, but in case it does then we have bigger problem at hand for production, why do we think current default value is good enough. And why did we provide it as GUC to control the value and set it to 5 secs, if can't even be used internally for testing?

Ah, you are right, I double-checked the GDD tests, we have already adjusted the tests long ago so they do not rely on the GDD period.

File: src/test/isolation2/sql/gdd/local-deadlock-03.sql [hunk 1]
```diff
@@ -1,3 +1,13 @@
+
+-- gdd can also detect local deadlocks, however it might break at
+-- different node with the local deadlock detector.  To make the local
+-- deadlock testcases stable we reset the gdd period to 2min so should
+-- not be triggered during the local deadlock tests.
+ALTER SYSTEM RESET gp_global_deadlock_detector_period;
+SELECT pg_reload_conf();
+-- start new session, which should always have newly reflected value
+1: SHOW gp_global_deadlock_detector_period;
+
```
Comment (pivotal-ning-yu) [inline src/test/isolation2/sql/gdd/local-deadlock-03.sql]: Cool, glad to know we have this magic.

Although we have only one `local-deadlock-*` test for now, we may have more in the future, so I still prefer to keep this preparation in a separate step.

**Comment by ashwinstar** [comment]
> On Aug 26, 2019, at 9:21 PM, Ning Yu <notifications@github.com> wrote:
> 
> @pivotal-ning-yu commented on this pull request.
> 
> In src/test/isolation2/sql/gdd/local-deadlock-03.sql:
> 
> > @@ -1,3 +1,13 @@
> +
> +-- gdd can also detect local deadlocks, however it might break at
> +-- different node with the local deadlock detector.  To make the local
> +-- deadlock testcases stable we reset the gdd period to 2min so should
> +-- not be triggered during the local deadlock tests.
> +ALTER SYSTEM RESET gp_global_deadlock_detector_period;
> +SELECT pg_reload_conf();
> +-- start new session, which should always have newly reflected value
> +1: SHOW gp_global_deadlock_detector_period;
> +
> Cool, glad to know we have this magic.
> 
> Although we have only one local-deadlock-* test for now, we may have more in the future, so I still prefer to keep this preparation in a separate step.
> 
Lets separate it when needed instead of having it separate and not using it. Like to code for current things than coding for imaginary future (only for on-disk change and all we need to be super future proof). For rest of things where cost of change is nothing, coding based on current requirements only is better.

**Comment by pivotal-ning-yu** [comment]
On Tue, Aug 27, 2019 at 12:33 PM Ashwin Agrawal <notifications@github.com>
wrote:

>
> > On Aug 26, 2019, at 9:21 PM, Ning Yu <notifications@github.com> wrote:
> >
> > @pivotal-ning-yu commented on this pull request.
> >
> > In src/test/isolation2/sql/gdd/local-deadlock-03.sql:
> >
> > > @@ -1,3 +1,13 @@
> > +
> > +-- gdd can also detect local deadlocks, however it might break at
> > +-- different node with the local deadlock detector. To make the local
> > +-- deadlock testcases stable we reset the gdd period to 2min so should
> > +-- not be triggered during the local deadlock tests.
> > +ALTER SYSTEM RESET gp_global_deadlock_detector_period;
> > +SELECT pg_reload_conf();
> > +-- start new session, which should always have newly reflected value
> > +1: SHOW gp_global_deadlock_detector_period;
> > +
> > Cool, glad to know we have this magic.
> >
> > Although we have only one local-deadlock-* test for now, we may have
> more in the future, so I still prefer to keep this preparation in a
> separate step.
> >
> Lets separate it when needed instead of having it separate and not using
> it. Like to code for current things than coding for imaginary future (only
> for on-disk change and all we need to be super future proof). For rest of
> things where cost of change is nothing, coding based on current
> requirements only is better.

Sure, makes sense.

**Pull request merged**